crates/agent-sql/src/publications.rs

```diff
@@ -208,6 +208,123 @@ pub async fn resolve_spec_rows(
     .await
 }
 
+#[derive(Debug)]
+pub struct Tenant {
+    pub name: String,
+    pub captures_quota: i32,
+    pub derivations_quota: i32,
+    pub materializations_quota: i32,
+    pub collections_quota: i32,
+    pub captures_used: i32,
+    pub derivations_used: i32,
+    pub materializations_used: i32,
+    pub collections_used: i32,
+}
+
+#[derive(Debug)]
+struct TenantFromDB {
+    pub name: String,
+    pub captures_quota: i32,
+    pub derivations_quota: i32,
+    pub materializations_quota: i32,
+    pub collections_quota: i32,
+    // These are Option<i64> because even if we add `coalesce(.., 0) to the query
+    // sqlx still infers them as nullable. So for our purposes assume None to be 0
+    pub captures_used: Option<i64>,
+    pub derivations_used: Option<i64>,
+    pub materializations_used: Option<i64>,
+    pub collections_used: Option<i64>,
+}
+
+impl From<TenantFromDB> for Tenant {
+    fn from(db: TenantFromDB) -> Self {
+        Tenant {
+            name: db.name,
+            captures_quota: db.captures_quota,
+            derivations_quota: db.derivations_quota,
+            materializations_quota: db.materializations_quota,
+            collections_quota: db.collections_quota,
+            captures_used: db.captures_used.unwrap_or(0) as i32,
+            derivations_used: db.derivations_used.unwrap_or(0) as i32,
+            materializations_used: db.materializations_used.unwrap_or(0) as i32,
+            collections_used: db.collections_used.unwrap_or(0).try_into().unwrap(),
+        }
+    }
+}
+
+pub async fn find_tenant_quotas(
+    tenant_names: Vec<String>,
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+) -> sqlx::Result<Vec<Tenant>> {
+    sqlx::query_as!(
+        TenantFromDB,
+        r#"
+        select
+            tenants.tenant as name,
+            tenants.captures_quota,
+            tenants.derivations_quota,
+            tenants.materializations_quota,
+            tenants.collections_quota,
+            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'capture') as captures_used,
+            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'collection' and live_specs.spec->'derivation' is not null) as derivations_used,
+            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'materialization') as materializations_used,
+            count(live_specs.catalog_name) filter (where live_specs.spec_type = 'collection') as collections_used
+        from tenants
+        join live_specs on live_specs.catalog_name LIKE tenants.tenant || '%' and (live_specs.spec->'shards'->>'disable')::boolean is not true
```

Comment:
supernit: use starts_with(live_specs.catalog_name, tenants.tenant) for consistency with how we do this everywhere else. This makes code search easier. _Probably_ there's no performance impact if postgres optimizes to this already.

Reply:
No action for now, but FYI: `spec` is stored a postgres `JSON` type, and not `JSONB`. It _must_ be this way because `JSONB` will re-order properties and that would break sops macs. But, in this application is means the query must actually parse all of these documents.

A way we can potentially improve this is by creating an index over the pre-extracted values used in this query. It would need to be confirmed, but postgres is smart about using a covering index to avoid needing to do the parsing work.

Reply:
Done